[css-anchor-position-1] Evaluate anchor() in calc()

anttijk opened the pull request:
#### 7240fde26436fed0bf903826c90f596c0207c5ae
<pre>
[css-anchor-position-1] Evaluate anchor() in calc()
<a href="https://bugs.webkit.org/show_bug.cgi?id=280328">https://bugs.webkit.org/show_bug.cgi?id=280328</a>
<a href="https://rdar.apple.com/problem/136659879">rdar://problem/136659879</a>

Reviewed by Sam Weinig.

calc(anchor(--foo)) evaluation support.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-query-fallback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-typed-om-expected.txt:
* Source/WebCore/css/CSSToLengthConversionData.cpp:
(WebCore::CSSToLengthConversionData::CSSToLengthConversionData):
* Source/WebCore/css/CSSToLengthConversionData.h:

Add Style::BuilderState to the conversion data. This is used for resolving anchor() functions.
It is only available when the conversion is invoked from Style::Builder but that is also the
time when anchor references must be resolved.

(WebCore::CSSToLengthConversionData::styleBuilderState const):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.h:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
(WebCore::CSSCalc::consumeAnchor):
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
(WebCore::CSSCalc::serializeMathFunctionArguments):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::computeInsetValue):
(WebCore::Style::AnchorPositionEvaluator::resolveAnchorValue):
(WebCore::Style::AnchorPositionEvaluator::evaluate):

Refactor a bit so when can share the code between the old CSSAnchorValue implementation and
the new calc() based implementation.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::BuilderState):
(WebCore::Style::BuilderState::cssPropertyID const):
* Source/WebCore/style/StyleResolver.h:

Canonical link: <a href="https://commits.webkit.org/284271@main">https://commits.webkit.org/284271@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa23b8f971150c7b4602daa47e9f6752b75551f0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68938 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21611 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73018 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20090 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71056 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56140 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19945 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13355 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72004 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/44138 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59521 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/35382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40803 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/16950 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18471 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62752 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17296 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74726 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12921 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/16542 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62540 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12959 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59604 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62443 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15294 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10421 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4029 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44141 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45216 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46412 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44957 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSToLengthConversionData.cpp:
```diff
@@ -38,13 +38,14 @@
+    , m_rootStyle(styleBuilderState.rootElementStyle())
+    , m_parentStyle(&styleBuilderState.parentStyle())
+    , m_renderView(styleBuilderState.document().renderView())
+    , m_elementForContainerUnitResolution(styleBuilderState.element())
```
small nit, but can we just name this builderState or state ? I think it's more consistent with the rest of the codebase.

anttijk replied:
CSSToLengthConversionData is not in Style namespace (though it should be), places outside the namespace generally use the prefix. See `Document::m_styleScope` for example.

anttijk replied:
I suppose shorter name works for locals.

weinig commented on Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
```diff
@@ -168,15 +169,18 @@ std::optional<double> evaluate(const IndirectNode<Hypot>& root, const Evaluation
-    if (!isValid && anchor->fallback)
+    auto result = Style::AnchorPositionEvaluator::evaluate(*options.conversionData->styleBuilderState(), *anchor);
+    if (!result) {
+        // FIXME: Invalid anchor without a valid fallback should make the declaration invalid at computed-value time.
```
Typo: FIXME -> FIMXE

;)

weinig commented on Source/WebCore/css/calc/CSSCalcTree+Evaluation.h:
```diff
@@ -31,6 +31,10 @@ namespace WebCore {
 class CSSCalcSymbolTable;
 
+namespace Style {
+class BuilderState;
```
This was probably from an earlier iteration where you passing this in explicitly instead of in the conversion data? Probably not needed now?

anttijk replied:
Yeah, tried a different approach first.

weinig commented on Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
```diff
@@ -627,7 +627,7 @@ static std::optional<TypedChild> consumeAnchor(CSSParserTokenRange& tokens, int
 
         if (tokens.peek().type() == PercentageToken)
-            return tokens.consumeIncludingWhitespace().numericValue();
+            return tokens.consumeIncludingWhitespace().numericValue() / 100;
```
Probably make things easier for future us if you throw the rest of the <anchor()> grammar at the top of the function:

```
<anchor-element> = <dashed-ident>
<anchor-side>    = inside | outside | top | left | right | bottom | start | end | self-start | self-end | <percentage> | center
```

Probably also worth adding a FIXME: stating this should be parsing the <percentage> properly, allowing for a nested calc() (or if that is not allowed by the spec, a comment explaining that, and we should get the spec updated to use [`<percentage-token>`](https://drafts.csswg.org/css-syntax-3/#typedef-percentage-token) instead).

anttijk replied:
Good point, nesting calc is allowed I believe.

weinig commented on Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
```diff
@@ -375,8 +375,8 @@ void serializeMathFunctionArguments(StringBuilder& builder, const IndirectNode<A
-        }, [&](double percentage) {
-            builder.append(percentage, '%');
+        }, [&](double value) {
+            builder.append(100 * value, '%');
```
I should have caught this in the last one, but this should probably use:

    formatCSSNumberValue(builder, value * 100, CSSPrimitiveValue::unitTypeString(CSSUnitType::CSS_PERCENTAGE));

as that will handle all the formatting edge cases (inf, lots of decimal places, etc).

darinadler replied:
I'm so glad you mentioned this! I noticed this a few hours ago looking over the last merged patch and was going to bring it up, but didn't get a chance before starting my work day.

anttijk replied:
That's a mouthful!

weinig commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -274,22 +274,20 @@ static LayoutRect computeAnchorRectRelativeToContainingBlock(CheckedRef<const Re
     if (anchorSideID == CSSValueCenter || anchorSideID == CSSValueInvalid) {
-        double percentage = 0.5;
-        if (anchorSideID != CSSValueCenter)
-            percentage = dynamicDowncast<CSSPrimitiveValue>(anchorValue.anchorSide())->valueDividingBy100IfPercentageDeprecated<double>();
```
Woo! One more `Deprecated` down!

anttijk committed:
[css-anchor-position-1] Evaluate anchor() in calc()
https://bugs.webkit.org/show_bug.cgi?id=280328
rdar://problem/136659879

Reviewed by Sam Weinig.

calc(anchor(--foo)) evaluation support.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-query-fallback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-typed-om-expected.txt:
* Source/WebCore/css/CSSToLengthConversionData.cpp:
(WebCore::CSSToLengthConversionData::CSSToLengthConversionData):
* Source/WebCore/css/CSSToLengthConversionData.h:

Add Style::BuilderState to the conversion data. This is used for resolving anchor() functions.
It is only available when the conversion is invoked from Style::Builder but that is also the
time when anchor references must be resolved.

(WebCore::CSSToLengthConversionData::styleBuilderState const):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.h:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
(WebCore::CSSCalc::consumeAnchor):
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
(WebCore::CSSCalc::serializeMathFunctionArguments):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::computeInsetValue):
(WebCore::Style::AnchorPositionEvaluator::resolveAnchorValue):
(WebCore::Style::AnchorPositionEvaluator::evaluate):

Refactor a bit so when can share the code between the old CSSAnchorValue implementation and
the new calc() based implementation.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::BuilderState):
(WebCore::Style::BuilderState::cssPropertyID const):
* Source/WebCore/style/StyleResolver.h:

Canonical link: https://commits.webkit.org/284271@main
